client/src/main/java/io/confluent/kafka/schemaregistry/client/rest/RestService.java

```diff
@@ -214,64 +241,131 @@ public RestService(String baseUrlConfig, boolean isForward) {
     setHttpConnectTimeoutMs(SchemaRegistryClientConfig.getHttpConnectTimeoutMs(configs));
     setHttpReadTimeoutMs(SchemaRegistryClientConfig.getHttpReadTimeoutMs(configs));
 
-    String basicCredentialsSource =
-        (String) configs.get(SchemaRegistryClientConfig.BASIC_AUTH_CREDENTIALS_SOURCE);
-    String bearerCredentialsSource =
-        (String) configs.get(SchemaRegistryClientConfig.BEARER_AUTH_CREDENTIALS_SOURCE);
+    String basicCredentialsSource = (String) configs.get(
+        SchemaRegistryClientConfig.BASIC_AUTH_CREDENTIALS_SOURCE);
+    String bearerCredentialsSource = (String) configs.get(
+        SchemaRegistryClientConfig.BEARER_AUTH_CREDENTIALS_SOURCE);
 
     if (isNonEmpty(basicCredentialsSource) && isNonEmpty(bearerCredentialsSource)) {
       throw new ConfigException(format(
           "Only one of '%s' and '%s' may be specified",
           SchemaRegistryClientConfig.BASIC_AUTH_CREDENTIALS_SOURCE,
-          SchemaRegistryClientConfig.BEARER_AUTH_CREDENTIALS_SOURCE
-      ));
+          SchemaRegistryClientConfig.BEARER_AUTH_CREDENTIALS_SOURCE));
 
     } else if (isNonEmpty(basicCredentialsSource)) {
-      BasicAuthCredentialProvider basicAuthCredentialProvider =
-          BasicAuthCredentialProviderFactory.getBasicAuthCredentialProvider(
+      BasicAuthCredentialProvider basicAuthCredentialProvider = BasicAuthCredentialProviderFactory
+          .getBasicAuthCredentialProvider(
               basicCredentialsSource,
-              configs
-          );
+              configs);
       setBasicAuthCredentialProvider(basicAuthCredentialProvider);
 
     } else if (isNonEmpty(bearerCredentialsSource)) {
       BearerAuthCredentialProvider bearerAuthCredentialProvider =
           BearerAuthCredentialProviderFactory.getBearerAuthCredentialProvider(
               bearerCredentialsSource,
-              configs
-          );
+              configs);
       setBearerAuthCredentialProvider(bearerAuthCredentialProvider);
     }
 
+    if (SchemaRegistryClientConfig.getUrlRandomize(configs)) {
+      baseUrls.randomizeIndex();
+    }
+
     String proxyHost = (String) configs.get(SchemaRegistryClientConfig.PROXY_HOST);
     Object proxyPortVal = configs.get(SchemaRegistryClientConfig.PROXY_PORT);
     Integer proxyPort = proxyPortVal instanceof String
-                        ? Integer.valueOf((String) proxyPortVal)
-                        : (Integer) proxyPortVal;
+        ? Integer.valueOf((String) proxyPortVal)
+        : (Integer) proxyPortVal;
 
     if (isValidProxyConfig(proxyHost, proxyPort)) {
       setProxy(proxyHost, proxyPort);
     }
+  }
 
-    if (SchemaRegistryClientConfig.getUrlRandomize(configs)) {
-      baseUrls.randomizeIndex();
+  HttpClient getApacheHttpClient() {
+    if (this.httpClient != null) {
+      return this.httpClient;
+    }
+    synchronized (this) {
+      if (this.useApacheHttpClient) {
+        this.httpClient = createNewHttpClient();
+      }
+    }
+    return this.httpClient;
```

Comment:
Can change to double-checked locking pattern (assumes this.httpClient is volatile)
```
   if (!this.useApacheHttpClient) {
       return null;
   }
   if (this.httpClient == null) {
    synchronized (this) {
      if (this.httpClient == null) {
        this.httpClient = createNewHttpClient();
      }
    }
    return this.httpClient;
```